dapp/src/node/blockchain.rs

```diff
@@ -0,0 +1,142 @@
+use std::collections::HashMap;
+
+use ethers::{
+    prelude::{
+        transaction::eip2718::TypedTransaction, Address, Block, NameOrAddress, Signer, Transaction,
+        TransactionReceipt, TxHash, H256, U256, U64,
+    },
+    utils::keccak256,
+};
+use evm_adapters::Evm;
+
+use super::{methods::EthResponse, SharedNode};
+
+#[derive(Default)]
+pub struct Blockchain {
+    pub(crate) blocks_by_number: HashMap<U64, Block<TxHash>>,
+    pub(crate) blocks_by_hash: HashMap<H256, Block<TxHash>>,
```

Comment:
`Vec` is also a good choice, since we will almost never need the `find` method on it. We will almost always fetch by block number, which is `blocks_by_number.get(&block_number)` (which should not be inefficient compared to `HashMap` lookups).